hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationFuture.java

```diff
@@ -99,21 +101,27 @@ protected E resolveAndThrowIfException(Object unresolved) throws ExecutionExcept
     }
 
     // public for tests
-    public static <T> T returnOrThrowWithGetConventions(Object resolved) throws ExecutionException, InterruptedException {
-        if (!(resolved instanceof ExceptionalResult)) {
-            return (T) resolved;
-        } else {
-            Throwable cause = ((ExceptionalResult) resolved).getCause();
-            if (cause instanceof CancellationException) {
-                throw (CancellationException) cause;
-            } else if (cause instanceof ExecutionException) {
-                throw (ExecutionException) cause;
-            } else if (cause instanceof InterruptedException) {
-                throw (InterruptedException) cause;
+    public static <T> T returnOrThrowWithGetConventions(Object response) throws ExecutionException, InterruptedException {
+        if (response instanceof ExceptionalResult) {
```

Comment:
The exception handling should still happen only inside the `if (response instanceof ExceptionalResult) {` block, instead of replacing the `response` value and working on that. For example if `response` is a `RuntimeException` instance (as a normal return value, not an exceptional response), with the previous code it would be returned as the normal return value but now it will be regarded as an exceptional outcome.

Take this as an example:
```
HazelcastInstance hz = Hazelcast.newHazelcastInstance();
IAtomicReference ref = hz.getCPSubsystem().getAtomicReference("test");
ref.set(new RuntimeException("boom!"));
System.out.println(ref.getAsync().toCompletableFuture().get());
```

With previous code it works properly, returning the `RuntimeException` instance, while with this code it throws
```
Exception in thread "main" java.util.concurrent.ExecutionException: java.lang.RuntimeException: boom!
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.returnOrThrowWithGetConventions(InvocationFuture.java:121)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:100)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:600)
	at Scratch.main(scratch_148.java:17)
Caused by: java.lang.RuntimeException: boom!
	at Scratch.main(scratch_148.java:16)
	at ------ submitted from ------.()
	at com.hazelcast.internal.util.ExceptionUtil.cloneExceptionWithFixedAsyncStackTrace(ExceptionUtil.java:266)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.returnOrThrowWithGetConventions(InvocationFuture.java:111)
	... 3 more
```

Reply:
Good catch. I have restored the behavior. The code in the ClientInvocationFuture misguided me, and it seems that it was wrong in the first place.